adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleBanner.m

```diff
@@ -0,0 +1,275 @@
+    _bannerSize = [self vungleAdSizeForAdSize:[adConfiguration adSize]];
+      
+    VungleAdNetworkExtras *networkExtras = adConfiguration.extras;
+    self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:adConfiguration.credentials.settings networkExtras:networkExtras];
+    self.uniquePubRequestID = [networkExtras.UUID copy];
+
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationBannerLoadCompletionHandler origAdLoadHandler = [completionHandler copy];
+    /// Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationBannerAdEventDelegate>(
+      id<GADMediationBannerAd> ad, NSError *error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationBannerAdEventDelegate> delegate = nil;
+      if (origAdLoadHandler) {
+        delegate = origAdLoadHandler(ad, error);
+      }
+      origAdLoadHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestBannerAd {
+  if (!IsGADAdSizeValid(_bannerSize)) {
+    NSString *errorMessage = [NSString stringWithFormat:@"Unsupported ad size requested for Vungle. Size: %@", NSStringFromGADAdSize(_bannerSize)];
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(GADMAdapterVungleErrorBannerSizeMismatch, errorMessage);
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+
+  if (!self.desiredPlacement.length) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(GADMAdapterVungleErrorInvalidServerParameters, @"Placement ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+
+  if (![[GADMAdapterVungleBiddingRouter sharedInstance] isSDKInitialized]) {
+    NSString *appID = [GADMAdapterVungleUtils findAppID:_adConfiguration.credentials.settings];
+    [[GADMAdapterVungleBiddingRouter sharedInstance] initWithAppId:appID delegate:self];
+    return;
+  }
+
+  [self loadAd];
+}
+
+- (GADAdSize)vungleAdSizeForAdSize:(GADAdSize)adSize {
+  // An array of supported ad sizes.
+  GADAdSize shortBannerSize = GADAdSizeFromCGSize(kVNGBannerShortSize);
+  NSArray<NSValue *> *potentials = @[
+    NSValueFromGADAdSize(GADAdSizeMediumRectangle), NSValueFromGADAdSize(GADAdSizeBanner),
+    NSValueFromGADAdSize(GADAdSizeLeaderboard), NSValueFromGADAdSize(shortBannerSize)
+  ];
+
+  GADAdSize closestSize = GADClosestValidSizeForAdSizes(adSize, potentials);
+  CGSize size = CGSizeFromGADAdSize(closestSize);
+  if (size.height == GADAdSizeBanner.size.height) {
+    if (size.width < GADAdSizeBanner.size.width) {
+      return shortBannerSize;
+    } else {
+      return GADAdSizeBanner;
+    }
+  } else if (size.height == GADAdSizeLeaderboard.size.height) {
+    return GADAdSizeLeaderboard;
+  } else if (size.height == GADAdSizeMediumRectangle.size.height) {
+    return GADAdSizeMediumRectangle;
+  }
+
+  return GADAdSizeInvalid;
+}
```

Comment:
We already have the `vungleAdSizeForAdSize` method in `GADMAdapterVungleBanner` class, so we can consider moving this to the utils class so it can be used by both waterfall and bidding requests.

Reply:
I moved it to the Utils file.